draft-ietf-mls-protocol.md

```diff
@@ -1101,44 +1102,78 @@ an explicit, application-defined identifier to a KeyPackage.
-the harm to group agreement properties, or whether there are alternative
-approaches to deniability that could be compatible with the other approach. -->
+struct {
+    HPKEPublicKey public_key;
+    opaque parent_hash<0..255>;
+    HPKEPublicKey original_child_resolution<0..2^32-1>;
+} ParentHashInput;
+~~~~~
+
+The Parent Hash of P with Co-Path Child S is the hash of a `ParentHashInput` object
+populated as follows. The field `public_key` contains the HPKE public key of P. If P
+is the root then `parent_hash` is set to the all-zeroes vector of size `Nh`.
+Otherwise `parent_hash` is the Parent Hash of P's parent with P's sibling as the
+co-path child.
+
+Finally, `original_child_resolution` is the array of `HPKEPublicKey` values of the
+nodes in the resolution of S but with the `unmerged_leaves` of P omitted. For
+example, in the ratchet tree depicted in {{resolution-example}} the
+`ParentHashInput` of node 5 with co-path child 4 would contain an empty
+`original_child_resolution` since 4's resolution includes only itself but 4 is also
+an unmerged leaf of 5. Meanwhile, the `ParentHashInput` of node 5 with co-path child
+6 has an array with one element in it; namely the HPKE public key of 6.
+
+### Using Parent Hashes
+
+The Parent Hash of P appears in three types of structs. If V is itself a parent node
+then the P's Parent Hash is stored in the `parent_hash` fields of both V's
+`ParentHashInput` struct and V's `ParentNode` struct. (The `ParentNode` struct is
+used to encapsulate all public information about V that must be conveyed to a new
+member joining the group as well as to define the Tree Hash of node V.)
+
+If, on the other hand, V is a leaf and its KeyPackage contains the `parent_hash`
+extension then the Parent Hash of P (with V's sibling as co-path child) is stored in
+that field. In particular, the extension MUST be present in the `leaf_key_package`
+field of an `UpdatePath` object. (This way, the signature of such a KeyPackage also
+serves to attest to which keys the group member introduced into the ratchet tree and
+to whom the corresponding secret keys were sent. This helps prevent malicious insiders
+from constructing artificial ratchet trees with a node V whose HPKE secret key is
+known to the insider yet where the insider isn't assigned a leaf in the subtree rooted
+at V. Indeed, such a ratchet tree would violate the tree invariant.)
+
+### Verifying Parent Hashes
+
+To this end, when processing a Commit message clients MUST recompute the
+expected value of `parent_hash` for the committer's new leaf and verify that it
+matches the `parent_hash` value in the supplied `leaf_key_package`. Moreover, when
+joining a group, new members MUST authenticate each non-blank parent node P. A parent
+node P is authenticated by checking that it has children V and S such that V's
+`ParentNode` struct has its `parent_hash` field set to the Parent Hash of P with
+Co-Path Child S.
```

Comment:
This section should address the problem noted in #435.  Namely, `V` should either be `left(P)` or `left^k(right(P)`, for the smallest possible `k >= 0`. Suggested text:

```suggestion
To this end, when processing a Commit message clients MUST recompute the
expected value of `parent_hash` for the committer's new leaf and verify that it
matches the `parent_hash` value in the supplied `leaf_key_package`. Moreover, when
joining a group, new members MUST authenticate each non-blank parent node P. A parent
node P is authenticated by performing the following check: 

* Let L and R be the left and right children of P, respectively
* If L.parent_hash is equal to the Parent Hash of P with Co-Path Child R, the check passes 
* If R is blank, replace R with its left child until R is either non-blank or a leaf node
* If R is a leaf node, the check fails
* If R.parent_hash is equal to the Parent Hash of P with Co-Path Child L, the check passes
* Otherwise, the check fails

The left-child recursion under the right child of P is necessary because the expansion of 
the tree to the right due to Add proposals can cause blank nodes to be interposed 
between a parent node and its right child. 
```

cc @MartaMularczyk @dajost

Reply:
Nice! Looks good to me.